File: ui/js/native/styles.js
// Fotox — live Layer Style dialogs (M6-T08/T09).
//
// Layer ▸ Layer Style ▸ Drop Shadow / Inner Shadow / Outer Glow / Color
// Overlay / Stroke and Blending Options edit the active layer. Every change
// sends the whole `LayerStyles` (`set_layer_style`); the engine merges the
// repeated edits into one history step. Cancel sends the styles back as they
// were.

import { openDialog } from "../dialogs.js";
import { toast } from "../tooltip.js";
import { activeLayerInfo, sendCommand } from "./layers-panel.js";
import { hexToRgba16, rgba16ToHex } from "./tools.js";
import { currentGradientResolved, gradientEditor, resolveSwatches } from "./gradients.js";
import { currentPattern, patternList } from "./patterns.js";

const blendId = (name) => String(name || "Normal").toLowerCase().replace(/[^a-z]+/g, "_").replace(/_add_?$/, "");
const blendLabel = (id) => String(id || "normal").split("_").map((w) => w[0].toUpperCase() + w.slice(1)).join(" ");

// dialog id → [styles field, defaults, to dialog values, from dialog values]
const EFFECTS = {
  "style-drop-shadow": {
    field: "drop_shadow",
    defaults: { enabled: true, blend: "multiply", color: [0, 0, 0, 65535], opacity: 0.75, angle: 120, use_global_light: true, distance: 5, spread: 0, size: 5 },
    toValues: (e) => ({ "Blend Mode:": blendLabel(e.blend), Colour: rgba16ToHex(e.color), "Opacity:": Math.round(e.opacity * 100), "Angle:": e.angle, "Use Global Light": e.use_global_light, "Distance:": e.distance, "Spread:": e.spread, "Size:": e.size }),
    fromValues: (v, e) => ({ ...e, blend: blendId(v["Blend Mode:"]), color: hexToRgba16(v.Colour), opacity: v["Opacity:"] / 100, angle: v["Angle:"], use_global_light: !!v["Use Global Light"], distance: v["Distance:"], spread: v["Spread:"], size: v["Size:"] }),
  },
  "style-inner-shadow": {
    field: "inner_shadow",
    defaults: { enabled: true, blend: "multiply", color: [0, 0, 0, 65535], opacity: 0.75, angle: 120, use_global_light: false, distance: 5, choke: 0, size: 5 },
    toValues: (e) => ({ "Blend Mode:": blendLabel(e.blend), Colour: rgba16ToHex(e.color), "Opacity:": Math.round(e.opacity * 100), "Angle:": e.angle, "Distance:": e.distance, "Choke:": e.choke, "Size:": e.size }),
    // FAST: the dialog has no Use Global Light box; its angle always applies.
    fromValues: (v, e) => ({ ...e, blend: blendId(v["Blend Mode:"]), color: hexToRgba16(v.Colour), opacity: v["Opacity:"] / 100, angle: v["Angle:"], use_global_light: false, distance: v["Distance:"], choke: v["Choke:"], size: v["Size:"] }),
  },
  "style-outer-glow": {
    field: "outer_glow",
    defaults: { enabled: true, blend: "screen", opacity: 0.75, color: [65535, 65535, 48830, 65535], spread: 0, size: 5 },
    toValues: (e) => ({ "Blend Mode:": blendLabel(e.blend), Colour: rgba16ToHex(e.color), "Opacity:": Math.round(e.opacity * 100), "Spread:": e.spread, "Size:": e.size }),
    fromValues: (v, e) => ({ ...e, blend: blendId(v["Blend Mode:"]), color: hexToRgba16(v.Colour), opacity: v["Opacity:"] / 100, spread: v["Spread:"], size: v["Size:"] }),
  },
  "style-color-overlay": {
    field: "color_overlay",
    defaults: { enabled: true, blend: "normal", color: [65535, 0, 0, 65535], opacity: 1 },
    toValues: (e) => ({ "Blend Mode:": blendLabel(e.blend), Colour: rgba16ToHex(e.color), "Opacity:": Math.round(e.opacity * 100) }),
    fromValues: (v, e) => ({ ...e, blend: blendId(v["Blend Mode:"]), color: hexToRgba16(v.Colour), opacity: v["Opacity:"] / 100 }),
  },
  "style-stroke": {
    field: "stroke",
    defaults: { enabled: true, size: 3, position: "outside", blend: "normal", opacity: 1, color: [0, 0, 0, 65535] },
    toValues: (e) => ({ Colour: rgba16ToHex(e.color), "Size:": e.size, "Position:": { outside: "Outside", inside: "Inside", center: "Centre" }[e.position], "Blend Mode:": blendLabel(e.blend), "Opacity:": Math.round(e.opacity * 100) }),
    fromValues: (v, e) => ({ ...e, color: hexToRgba16(v.Colour), size: Math.max(0, v["Size:"]), position: { Outside: "outside", Inside: "inside", Centre: "center" }[v["Position:"]] || "outside", blend: blendId(v["Blend Mode:"]), opacity: v["Opacity:"] / 100 }),
  },
  // M12-T04.
  "style-inner-glow": {
    field: "inner_glow",
    defaults: { enabled: true, blend: "screen", opacity: 0.75, color: [65535, 65535, 48830, 65535], choke: 0, size: 5 },
    toValues: (e) => ({ "Blend Mode:": blendLabel(e.blend), Colour: rgba16ToHex(e.color), "Opacity:": Math.round(e.opacity * 100), "Choke:": e.choke, "Size:": e.size }),
    fromValues: (v, e) => ({ ...e, blend: blendId(v["Blend Mode:"]), color: hexToRgba16(v.Colour), opacity: v["Opacity:"] / 100, choke: v["Choke:"], size: v["Size:"] }),
  },
  "style-satin": {
    field: "satin",
    defaults: { enabled: true, blend: "multiply", color: [0, 0, 0, 65535], opacity: 0.5, angle: 19, distance: 11, size: 14, invert: true },
    toValues: (e) => ({ "Blend Mode:": blendLabel(e.blend), Colour: rgba16ToHex(e.color), "Opacity:": Math.round(e.opacity * 100), "Angle:": e.angle, "Distance:": e.distance, "Size:": e.size, Invert: e.invert }),
    fromValues: (v, e) => ({ ...e, blend: blendId(v["Blend Mode:"]), color: hexToRgba16(v.Colour), opacity: v["Opacity:"] / 100, angle: v["Angle:"], distance: v["Distance:"], size: v["Size:"], invert: !!v.Invert }),
  },
  "style-bevel": {
    field: "bevel",
    defaults: {
      enabled: true, style: "inner_bevel", depth: 100, up: true, size: 5, soften: 0, angle: 120, use_global_light: true, altitude: 30,
      highlight_blend: "screen", highlight_color: [65535, 65535, 65535, 65535], highlight_opacity: 0.75,
      shadow_blend: "multiply", shadow_color: [0, 0, 0, 65535], shadow_opacity: 0.75,
    },
    toValues: (e) => ({
      "Style:": { inner_bevel: "Inner Bevel", outer_bevel: "Outer Bevel", emboss: "Emboss", pillow_emboss: "Pillow Emboss" }[e.style] || "Inner Bevel",
      "Depth:": e.depth, "Direction:": e.up ? "Up" : "Down", "Size:": e.size, "Soften:": e.soften, "Angle:": e.angle, "Altitude:": e.altitude,
      "Highlight Mode:": blendLabel(e.highlight_blend), Highlight: rgba16ToHex(e.highlight_color), "Highlight Opacity:": Math.round(e.highlight_opacity * 100),
      "Shadow Mode:": blendLabel(e.shadow_blend), Shadow: rgba16ToHex(e.shadow_color), "Shadow Opacity:": Math.round(e.shadow_opacity * 100),
    }),
    fromValues: (v, e) => ({
      ...e,
      // FAST: Stroke Emboss acts as Emboss.
      style: { "Inner Bevel": "inner_bevel", "Outer Bevel": "outer_bevel", Emboss: "emboss", "Pillow Emboss": "pillow_emboss", "Stroke Emboss": "emboss" }[v["Style:"]] || "inner_bevel",
      depth: v["Depth:"], up: v["Direction:"] !== "Down" && v["Direction:"] !== 1, size: v["Size:"], soften: v["Soften:"],
      angle: v["Angle:"], use_global_light: false, altitude: v["Altitude:"],
      highlight_blend: blendId(v["Highlight Mode:"]), highlight_color: hexToRgba16(v.Highlight), highlight_opacity: v["Highlight Opacity:"] / 100,
      shadow_blend: blendId(v["Shadow Mode:"]), shadow_color: hexToRgba16(v.Shadow), shadow_opacity: v["Shadow Opacity:"] / 100,
    }),
  },
};

/** Whether the engine implements this dialog id (without the `dlg:`). */
export function isStyleDialog(id) {
  return id in EFFECTS || id === "blending-options" || id === "style-gradient-overlay" || id === "style-pattern-overlay";
}

/** Open a live style dialog for the active layer. */
export function openStyleDialog(id) {
  const layer = activeLayerInfo();
  if (!layer) { toast("Select a layer first"); return; }
  if (layer.kind === "group" || layer.kind === "adjustment") { toast("Layer styles need a pixel, shape, text or fill layer"); return; }
  if (id === "blending-options") { blendingOptions(layer); return; }
  if (id === "style-gradient-overlay") { gradientOverlay(layer); return; }
  if (id === "style-pattern-overlay") { patternOverlay(layer); return; }
  const spec = EFFECTS[id];
  const original = layer.styles || null;
  const effect = (original && original[spec.field]) || spec.defaults;
  const set = (styles) => sendCommand({ op: "set_layer_style", layer: { id: layer.id }, styles });
  const withEffect = (values) => ({ ...(original || {}), [spec.field]: spec.fromValues(values, effect) });
  openDialog(id, {
    title: `Layer Style — ${layer.name}`,
    values: spec.toValues(effect),
    onChange: (values) => set(withEffect(values)),
    onOk: (values) => set(withEffect(values)),
    onCancel: () => set(original),
  });
  // Show the effect at once, as Photoshop does when the dialog opens.
  set(withEffect(spec.toValues(effect)));
}

function blendingOptions(layer) {
  const original = { opacity: layer.opacity, fill: layer.fill, blend: layer.blend };
  const set = (props) => sendCommand({ op: "set_layer_props", layer: { id: layer.id }, props });
  const from = (v) => ({ opacity: Math.min(1, Math.max(0, v["Opacity:"] / 100)), fill: Math.min(1, Math.max(0, v["Fill Opacity:"] / 100)), blend: blendId(v["Blend Mode:"]) });
  openDialog("blending-options", {
    title: `Layer Style — ${layer.name}`,
    values: { "Blend Mode:": blendLabel(layer.blend), "Opacity:": Math.round(layer.opacity * 100), "Fill Opacity:": Math.round(layer.fill * 100) },
    onChange: (v) => set(from(v)),
    onOk: (v) => set(from(v)),
    onCancel: () => set(original),
  });
}

const GRADIENT_STYLES = ["Linear", "Radial", "Angle", "Reflected", "Diamond"];

/** Gradient Overlay (M12-T04). FAST: applied on OK, no live update. */
function gradientOverlay(layer) {
  const original = layer.styles || null;
  const old = original?.gradient_overlay;
  const work = old ? structuredClone(old.gradient.gradient) : currentGradientResolved();
  openDialog("style-gradient-overlay", {
    title: `Layer Style — ${layer.name}`,
    width: 460,
    fields: [
      { type: "blend", mode: blendLabel(old?.blend || "normal") },
      { type: "range", label: "Opacity:", value: Math.round((old?.opacity ?? 1) * 100), min: 0, max: 100 },
      { type: "element", el: gradientEditor(work) },
      { type: "select", label: "Style:", options: GRADIENT_STYLES, value: old ? old.gradient.kind[0].toUpperCase() + old.gradient.kind.slice(1) : "Linear" },
      { type: "num", label: "Angle:", value: old?.gradient.angle ?? 90, unit: "°", w: 60 },
      { type: "num", label: "Scale:", value: old?.gradient.scale ?? 100, unit: "%", w: 60 },
      { type: "check", label: "Reverse", value: old?.gradient.reverse ?? false },
    ],
    onOk: (v) => {
      resolveSwatches(work);
      const effect = {
        enabled: true, blend: blendId(v["Blend Mode:"]), opacity: (Number(v["Opacity:"]) || 0) / 100,
        gradient: {
          gradient: work, kind: String(v["Style:"] || "Linear").toLowerCase(), angle: Number(v["Angle:"]) || 0,
          scale: Math.min(1000, Math.max(1, Number(v["Scale:"]) || 100)), reverse: !!v.Reverse, dither: true, offset: [0, 0],
        },
      };
      sendCommand({ op: "set_layer_style", layer: { id: layer.id }, styles: { ...(original || {}), gradient_overlay: effect } });
    },
  });
}

/** Pattern Overlay (M12-T04). FAST: applied on OK. */
function patternOverlay(layer) {
  const original = layer.styles || null;
  const old = original?.pattern_overlay;
  const selected = { id: old ? old.pattern : currentPattern() };
  openDialog("style-pattern-overlay", {
    title: `Layer Style — ${layer.name}`,
    fields: [
      { type: "blend", mode: blendLabel(old?.blend || "normal") },
      { type: "range", label: "Opacity:", value: Math.round((old?.opacity ?? 1) * 100), min: 0, max: 100 },
      { type: "element", el: patternList(selected) },
      { type: "num", label: "Scale:", value: old?.scale ?? 100, unit: "%", w: 60 },
    ],
    onOk: (v) => {
      if (selected.id == null) { toast("Pick a pattern (Edit ▸ Define Pattern first)"); return; }
      const effect = { enabled: true, blend: blendId(v["Blend Mode:"]), opacity: (Number(v["Opacity:"]) || 0) / 100, pattern: selected.id, scale: Math.min(1000, Math.max(1, Number(v["Scale:"]) || 100)) };
      sendCommand({ op: "set_layer_style", layer: { id: layer.id }, styles: { ...(original || {}), pattern_overlay: effect } });
    },
  });
}

File: crates/fx-core/src/styles.rs
//! Layer styles (M6-T08, D-054): Drop Shadow, Outer Glow, Inner Shadow,
//! Color Overlay and Stroke; M12-T04 adds Inner Glow, Satin, Bevel & Emboss
//! (a shadow and a highlight pass), Gradient Overlay and Pattern Overlay.
//!
//! The parameters are the truth; each effect's pixels are a derived tile
//! cache on the layer ([`crate::Layer::effects`]), drawn by the engine from
//! the layer's alpha at the level being shown, like a shape's tiles.
//!
//! Lengths are document pixels at level 0; angles are degrees, Photoshop's
//! convention (0° = light from the right, 90° = from above, so the shadow
//! falls down).

use serde::{Deserialize, Serialize};

use crate::blend::BlendMode;

/// The effects, in the order they are composited (bottom → top).
/// VERIFY: Photoshop's exact stacking of the interior effects.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EffectKind {
	DropShadow,
	OuterGlow,
	PatternOverlay,
	GradientOverlay,
	ColorOverlay,
	Satin,
	InnerGlow,
	InnerShadow,
	Stroke,
	BevelShadow,
	BevelHighlight,
}

impl EffectKind {
	pub const ALL: [EffectKind; 11] = [
		EffectKind::DropShadow,
		EffectKind::OuterGlow,
		EffectKind::PatternOverlay,
		EffectKind::GradientOverlay,
		EffectKind::ColorOverlay,
		EffectKind::Satin,
		EffectKind::InnerGlow,
		EffectKind::InnerShadow,
		EffectKind::Stroke,
		EffectKind::BevelShadow,
		EffectKind::BevelHighlight,
	];

	/// Index into [`crate::Layer::effects`].
	pub fn index(self) -> usize {
		self as usize
	}

	pub fn from_index(i: usize) -> Option<Self> {
		Self::ALL.get(i).copied()
	}

	/// Drawn under the layer's content (the rest are drawn over it).
	pub fn below_content(self) -> bool {
		matches!(self, EffectKind::DropShadow | EffectKind::OuterGlow)
	}
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StrokePosition {
	Outside,
	Inside,
	Center,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct DropShadow {
	pub enabled: bool,
	pub blend: BlendMode,
	pub color: [u16; 4],
	pub opacity: f32,
	pub angle: f64,
	pub use_global_light: bool,
	pub distance: f64,
	/// Percent of `size` that is dilation instead of blur (Photoshop's Spread).
	pub spread: f64,
	pub size: f64,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct OuterGlow {
	pub enabled: bool,
	pub blend: BlendMode,
	pub opacity: f32,
	pub color: [u16; 4],
	pub spread: f64,
	pub size: f64,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct InnerShadow {
	pub enabled: bool,
	pub blend: BlendMode,
	pub color: [u16; 4],
	pub opacity: f32,
	pub angle: f64,
	pub use_global_light: bool,
	pub distance: f64,
	/// Percent of `size` that is erosion instead of blur (Photoshop's Choke).
	pub choke: f64,
	pub size: f64,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct ColorOverlay {
	pub enabled: bool,
	pub blend: BlendMode,
	pub color: [u16; 4],
	pub opacity: f32,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Stroke {
	pub enabled: bool,
	pub size: f64,
	pub position: StrokePosition,
	pub blend: BlendMode,
	pub opacity: f32,
	pub color: [u16; 4],
}

/// Bevel & Emboss styles (M12-T04).
#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BevelStyle {
	InnerBevel,
	OuterBevel,
	Emboss,
	PillowEmboss,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct BevelEmboss {
	pub enabled: bool,
	pub style: BevelStyle,
	/// Percent (1..=1000).
	pub depth: f64,
	/// Direction Up (true) / Down.
	pub up: bool,
	pub size: f64,
	pub soften: f64,
	pub angle: f64,
	pub use_global_light: bool,
	/// Degrees above the horizon.
	pub altitude: f64,
	pub highlight_blend: BlendMode,
	pub highlight_color: [u16; 4],
	pub highlight_opacity: f32,
	pub shadow_blend: BlendMode,
	pub shadow_color: [u16; 4],
	pub shadow_opacity: f32,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct InnerGlow {
	pub enabled: bool,
	pub blend: BlendMode,
	pub opacity: f32,
	pub color: [u16; 4],
	pub choke: f64,
	pub size: f64,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Satin {
	pub enabled: bool,
	pub blend: BlendMode,
	pub color: [u16; 4],
	pub opacity: f32,
	pub angle: f64,
	pub distance: f64,
	pub size: f64,
	pub invert: bool,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct GradientOverlay {
	pub enabled: bool,
	pub blend: BlendMode,
	pub opacity: f32,
	/// The gradient, angle, scale… placed over the canvas (FAST: Photoshop
	/// aligns it with the layer's bounds by default).
	pub gradient: crate::gradient::GradientLayer,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct PatternOverlay {
	pub enabled: bool,
	pub blend: BlendMode,
	pub opacity: f32,
	/// A document pattern's id (`Document::patterns`).
	pub pattern: u64,
	/// Percent.
	pub scale: f64,
}

impl Default for BevelEmboss {
	fn default() -> Self {
		Self {
			enabled: true,
			style: BevelStyle::InnerBevel,
			depth: 100.0,
			up: true,
			size: 5.0,
			soften: 0.0,
			angle: 120.0,
			use_global_light: true,
			altitude: 30.0,
			highlight_blend: BlendMode::Screen,
			highlight_color: [65_535; 4],
			highlight_opacity: 0.75,
			shadow_blend: BlendMode::Multiply,
			shadow_color: [0, 0, 0, 65_535],
			shadow_opacity: 0.75,
		}
	}
}

impl Default for InnerGlow {
	fn default() -> Self {
		Self {
			enabled: true,
			blend: BlendMode::Screen,
			opacity: 0.75,
			color: [65_535, 65_535, 48_830, 65_535],
			choke: 0.0,
			size: 5.0,
		}
	}
}

impl Default for Satin {
	fn default() -> Self {
		Self {
			enabled: true,
			blend: BlendMode::Multiply,
			color: [0, 0, 0, 65_535],
			opacity: 0.5,
			angle: 19.0,
			distance: 11.0,
			size: 14.0,
			invert: true,
		}
	}
}

impl Default for DropShadow {
	fn default() -> Self {
		Self {
			enabled: true,
			blend: BlendMode::Multiply,
			color: [0, 0, 0, 65_535],
			opacity: 0.75,
			angle: 120.0,
			use_global_light: true,
			distance: 5.0,
			spread: 0.0,
			size: 5.0,
		}
	}
}

impl Default for OuterGlow {
	fn default() -> Self {
		Self {
			enabled: true,
			blend: BlendMode::Screen,
			opacity: 0.75,
			color: [65_535, 65_535, 48_830, 65_535],
			spread: 0.0,
			size: 5.0,
		}
	}
}

impl Default for InnerShadow {
	fn default() -> Self {
		Self {
			enabled: true,
			blend: BlendMode::Multiply,
			color: [0, 0, 0, 65_535],
			opacity: 0.75,
			angle: 120.0,
			use_global_light: true,
			distance: 5.0,
			choke: 0.0,
			size: 5.0,
		}
	}
}

impl Default for ColorOverlay {
	fn default() -> Self {
		Self {
			enabled: true,
			blend: BlendMode::Normal,
			color: [65_535, 0, 0, 65_535],
			opacity: 1.0,
		}
	}
}

impl Default for Stroke {
	fn default() -> Self {
		Self {
			enabled: true,
			size: 3.0,
			position: StrokePosition::Outside,
			blend: BlendMode::Normal,
			opacity: 1.0,
			color: [0, 0, 0, 65_535],
		}
	}
}

/// A layer's styles: each effect is present (with `enabled` for the panel's
/// eye) or absent.
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct LayerStyles {
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub drop_shadow: Option<DropShadow>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub outer_glow: Option<OuterGlow>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub inner_shadow: Option<InnerShadow>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub color_overlay: Option<ColorOverlay>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub stroke: Option<Stroke>,
	// M12-T04.
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub bevel: Option<BevelEmboss>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub inner_glow: Option<InnerGlow>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub satin: Option<Satin>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub gradient_overlay: Option<GradientOverlay>,
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub pattern_overlay: Option<PatternOverlay>,
}

/// What an M12 effect needs beyond a colour and a coverage.
#[derive(Clone, Debug, PartialEq, Default)]
pub enum EffectExtra {
	#[default]
	None,
	/// Satin: the shift of the two copies (document pixels) and Invert.
	Satin {
		offset: (f64, f64),
		invert: bool,
	},
	/// Bevel & Emboss: the pass (highlight or shadow) of this style.
	Bevel {
		style: BevelStyle,
		depth: f64,
		up: bool,
		size: f64,
		soften: f64,
		/// Light direction: azimuth and altitude, radians.
		light: (f64, f64),
		highlight: bool,
	},
	Gradient(crate::gradient::GradientLayer),
	Pattern {
		id: u64,
		scale: f64,
	},
}

/// The parameters of one enabled effect, resolved for drawing.
#[derive(Clone, Debug, PartialEq)]
pub struct EffectParams {
	pub kind: EffectKind,
	pub blend: BlendMode,
	pub opacity: f32,
	pub color: [u16; 4],
	/// Shift of the source alpha, document pixels (shadows).
	pub offset: (f64, f64),
	/// Dilation (> 0) or erosion (< 0) radius before the blur.
	pub morph: f64,
	/// Blur radius (Photoshop's "Size" minus the spread part).
	pub blur: f64,
	pub stroke: Option<(f64, StrokePosition)>,
	pub extra: EffectExtra,
}

impl LayerStyles {
	/// Whether any effect would draw.
	pub fn any_enabled(&self) -> bool {
		EffectKind::ALL.into_iter().any(|kind| self.effect(kind, 120.0).is_some())
	}

	/// The enabled effect `kind`, resolved with the document's global light.
	pub fn effect(&self, kind: EffectKind, global_light: f64) -> Option<EffectParams> {
		let offset = |angle: f64, global: bool, distance: f64| {
			let a = if global { global_light } else { angle }.to_radians();
			// Light from `a`: the shadow is cast the other way (y grows down).
			(-a.cos() * distance, a.sin() * distance)
		};
		match kind {
			EffectKind::DropShadow => self.drop_shadow.as_ref().filter(|e| e.enabled).map(|e| EffectParams {
				kind,
				blend: e.blend,
				opacity: e.opacity,
				color: e.color,
				offset: offset(e.angle, e.use_global_light, e.distance),
				morph: e.size * e.spread.clamp(0.0, 100.0) / 100.0,
				blur: e.size * (1.0 - e.spread.clamp(0.0, 100.0) / 100.0),
				stroke: None,
				extra: EffectExtra::None,
			}),
			EffectKind::OuterGlow => self.outer_glow.as_ref().filter(|e| e.enabled).map(|e| EffectParams {
				kind,
				blend: e.blend,
				opacity: e.opacity,
				color: e.color,
				offset: (0.0, 0.0),
				morph: e.size * e.spread.clamp(0.0, 100.0) / 100.0,
				blur: e.size * (1.0 - e.spread.clamp(0.0, 100.0) / 100.0),
				stroke: None,
				extra: EffectExtra::None,
			}),
			EffectKind::InnerShadow => self.inner_shadow.as_ref().filter(|e| e.enabled).map(|e| EffectParams {
				kind,
				blend: e.blend,
				opacity: e.opacity,
				color: e.color,
				offset: offset(e.angle, e.use_global_light, e.distance),
				morph: e.size * e.choke.clamp(0.0, 100.0) / 100.0,
				blur: e.size * (1.0 - e.choke.clamp(0.0, 100.0) / 100.0),
				stroke: None,
				extra: EffectExtra::None,
			}),
			EffectKind::ColorOverlay => self.color_overlay.as_ref().filter(|e| e.enabled).map(|e| EffectParams {
				kind,
				blend: e.blend,
				opacity: e.opacity,
				color: e.color,
				offset: (0.0, 0.0),
				morph: 0.0,
				blur: 0.0,
				stroke: None,
				extra: EffectExtra::None,
			}),
			EffectKind::Stroke => self.stroke.as_ref().filter(|e| e.enabled && e.size > 0.0).map(|e| EffectParams {
				kind,
				blend: e.blend,
				opacity: e.opacity,
				color: e.color,
				offset: (0.0, 0.0),
				morph: 0.0,
				blur: 0.0,
				stroke: Some((e.size, e.position)),
				extra: EffectExtra::None,
			}),
			EffectKind::InnerGlow => self.inner_glow.as_ref().filter(|e| e.enabled).map(|e| EffectParams {
				kind,
				blend: e.blend,
				opacity: e.opacity,
				color: e.color,
				offset: (0.0, 0.0),
				morph: e.size * e.choke.clamp(0.0, 100.0) / 100.0,
				blur: e.size * (1.0 - e.choke.clamp(0.0, 100.0) / 100.0),
				stroke: None,
				extra: EffectExtra::None,
			}),
			EffectKind::Satin => self.satin.as_ref().filter(|e| e.enabled).map(|e| {
				let a = e.angle.to_radians();
				EffectParams {
					kind,
					blend: e.blend,
					opacity: e.opacity,
					color: e.color,
					offset: (0.0, 0.0),
					morph: 0.0,
					blur: e.size,
					stroke: None,
					extra: EffectExtra::Satin {
						offset: (a.cos() * e.distance, -a.sin() * e.distance),
						invert: e.invert,
					},
				}
			}),
			EffectKind::BevelShadow | EffectKind::BevelHighlight => self.bevel.as_ref().filter(|e| e.enabled && e.size > 0.0).map(|e| {
				let highlight = kind == EffectKind::BevelHighlight;
				let angle = if e.use_global_light { global_light } else { e.angle };
				EffectParams {
					kind,
					blend: if highlight { e.highlight_blend } else { e.shadow_blend },
					opacity: if highlight { e.highlight_opacity } else { e.shadow_opacity },
					color: if highlight { e.highlight_color } else { e.shadow_color },
					offset: (0.0, 0.0),
					morph: 0.0,
					blur: e.soften,
					stroke: None,
					extra: EffectExtra::Bevel {
						style: e.style,
						depth: e.depth,
						up: e.up,
						size: e.size,
						soften: e.soften,
						light: (angle.to_radians(), e.altitude.clamp(0.0, 90.0).to_radians()),
						highlight,
					},
				}
			}),
			EffectKind::GradientOverlay => self.gradient_overlay.as_ref().filter(|e| e.enabled).map(|e| EffectParams {
				kind,
				blend: e.blend,
				opacity: e.opacity,
				color: [65_535; 4],
				offset: (0.0, 0.0),
				morph: 0.0,
				blur: 0.0,
				stroke: None,
				extra: EffectExtra::Gradient(e.gradient.clone()),
			}),
			EffectKind::PatternOverlay => self.pattern_overlay.as_ref().filter(|e| e.enabled).map(|e| EffectParams {
				kind,
				blend: e.blend,
				opacity: e.opacity,
				color: [65_535; 4],
				offset: (0.0, 0.0),
				morph: 0.0,
				blur: 0.0,
				stroke: None,
				extra: EffectExtra::Pattern { id: e.pattern, scale: e.scale },
			}),
		}
	}

	/// How far outside a tile an effect reads the layer's alpha, level 0.
	pub fn reach(params: &EffectParams) -> f64 {
		let stroke = params.stroke.map_or(0.0, |(size, _)| size);
		let extra = match &params.extra {
			EffectExtra::Satin { offset, .. } => offset.0.abs().max(offset.1.abs()),
			EffectExtra::Bevel { size, .. } => size + 2.0,
			_ => 0.0,
		};
		params.offset.0.abs().max(params.offset.1.abs()) + params.morph.abs() + params.blur * 1.5 + stroke + extra + 2.0
	}
}
